src/metabase/sync/fetch_metadata.clj

```diff
@@ -20,13 +22,31 @@
    table    :- i/TableInstance]
   (driver/describe-table (driver.u/database->driver database) database table))
 
-(mu/defn fk-metadata :- i/FKMetadata
+(mu/defn fk-metadata
+  "Effectively a wrapper for [[metabase.driver/describe-fks]] that also validates the output against the schema."
+  [database :- i/DatabaseInstance & {:as args}]
+  (cond->> (driver/describe-fks (driver.u/database->driver database) database args)
+    ;; Validate the output against the schema, except in prod.
+    ;; This is a workaround for the fact that [[mu/defn]] can't check reducible collections yet
+    (not config/is-prod?)
+    (eduction (map #(mu.fn/validate-output {} i/FKMetadataEntry %)))))
+
+(mu/defn table-fk-metadata :- [:maybe [:sequential i/FKMetadataEntry]]
```

Comment:
I find the need for this function unnecessary, we can use `fk-metadata` with a `table-names` arg for this use case right?

Reply:
ah I just realized that initially `fk-metadata` is per-table and now you changed it to per-db.

But I think it's cleaner to makes `fk-metadata` deals with the fact that there is 2 ways to get table metadata.

I think you can make `fk-metadata` smarter and remove this function.

```clojure
(mu/defn fk-metadata
  "Effectively a wrapper for [[metabase.driver/describe-fks]] that also validates the output against the schema."
  [database :- i/DatabaseInstance & {:keys [table-names] :as args}]
  (let [driver (driver.u/database->driver database)]
    (cond->> (if (driver/database-supports? driver :describe-fks database)
               (driver/describe-fks (driver.u/database->driver database) database args)
               (flatten (for [table (or table-names (sync.u/db->sync-tables database))]
                          (vec (for [x (driver/describe-table-fks driver database table)]
                                 {:fk-table-name   (:name table)
                                  :fk-table-schema (:schema table)
                                  :fk-column-name  (:fk-column-name x)
                                  :pk-table-name   (:name (:dest-table x))
                                  :pk-table-schema (:schema (:dest-table x))
                                  :pk-column-name  (:dest-column-name x)})))))
      ;; Validate the output against the schema, except in prod.
      ;; This is a workaround for the fact that [[mu/defn]] can't check reducible collections yet
      (not config/is-prod?)
      (eduction (map #(mu.fn/validate-output {} i/FKMetadataEntry %))))))
```
For case where only fk-metadata for one table is needed, you need to pass in the name of the tables, otherwise it'll return all tables within that database.

This way, `sync-fks!` doesn't need to do the feature check at all.

Reply:
Seems like a nice idea, the problem is `fk-metadata` needs to take `:schema-names`, and I don't want to complicate this function further by implementing support for that if we don't need it yet.

But this kind of refactor will be possible in the future if we implement `driver/describe-fks` in terms of `driver/describe-table-fks`, which I'm planning to do.